### ISSUE TITLE
TestMultiConnectToSingleBrowser is flaky

### ISSUE THREAD

**Issue by inancgumus** [grafana/xk6-browser#861]
Investigate why the test fails to create the second page.

```
time="2023-04-26T06:02:15Z" level=error msg="communicating with browser: websocket: close 1006 (abnormal closure): unexpected EOF" category=cdp elapsed="0 ms" goroutine=983
time="2023-04-26T06:02:15Z" level=error msg="communicating with browser: websocket: close 1006 (abnormal closure): unexpected EOF" category=cdp elapsed="0 ms" goroutine=1165

--- FAIL: TestMultiConnectToSingleBrowser (0.19s)
    browser_test.go:249: 
        	Error Trace:	/home/runner/work/xk6-browser/xk6-browser/tests/browser_test.go:249
        	Error:      	Received unexpected error:
        	            	creating new page in browser context: creating a new blank page: Session with given id not found. (-32001)
        	Test:       	TestMultiConnectToSingleBrowser
        	Messages:   	failed to create page #2

time="2023-04-26T06:02:15Z" level=error msg="communicating with browser: websocket: close 1006 (abnormal closure): unexpected EOF" category=cdp elapsed="0 ms" goroutine=1299
time="2023-04-26T06:02:30Z" level=error msg="Failed to load resource: the server responded with a status of 401 (Unauthorized)" browser_source=network line_number=0 source=browser stacktrace="<nil>" url="http://127.0.0.1:34171/basic-auth/validuser/validpass"
time="2023-04-26T06:02:30Z" level=warning msg="url:http://127.0.0.1:34171/basic-auth/validuser/validpass method:GET err:fetching response body: context canceled" category="Response:bodySize:fetchBody" elapsed="0 ms" goroutine=8069
time="2023-04-26T06:02:30Z" level=warning msg="url:http://127.0.0.1:35603/static/empty.html method:GET err:fetching response body: context canceled" category="Response:bodySize:fetchBody" elapsed="0 ms" goroutine=8365
time="2023-04-26T06:02:32Z" level=warning msg="The specified value \"invalid date\" does not conform to the required format, \"yyyy-MM-dd\"." browser_source=rendering line_number=457 source=browser url=__xk6_browser_evaluation_script__
time="2023-04-26T06:02:34Z" level=error msg="communicating with browser: websocket: close 1006 (abnormal closure): unexpected EOF" category=cdp elapsed="0 ms" goroutine=10587
time="2023-04-26T06:02:34Z" level=error msg="process with PID 9495 unexpectedly ended: signal: killed" category=browser elapsed="18 ms" goroutine=10605
time="2023-04-26T06:02:36Z" level=error msg="process with PID 9918 unexpectedly ended: exit status 1" category=browser elapsed="0 ms" goroutine=11781
time="2023-04-26T06:02:37Z" level=error msg="communicating with browser: websocket: close 1006 (abnormal closure): unexpected EOF" category=cdp elapsed="0 ms" goroutine=12075
time="2023-04-26T06:02:37Z" level=error msg="process with PID 10081 unexpectedly ended: signal: killed" category=browser elapsed="3 ms" goroutine=1[197](https://github.com/grafana/xk6-browser/actions/runs/4805402538/jobs/8551776054#step:4:198)4
time="[202](https://github.com/grafana/xk6-browser/actions/runs/4805402538/jobs/8551776054#step:4:203)3-04-26T06:02:38Z" level=info msg="\"calling setTimeout...\"" source=browser-console-api
time="2023-04-26T06:02:39Z" level=info msg="\"creating element...\"" source=browser-console-api
time="2023-04-26T06:02:42Z" level=info msg="\"hello\"" source=browser-console-api
```

**Comment by ankur22** [comment]
While working on a [PR](https://github.com/grafana/xk6-browser/pull/929), I found that this test was flakey. Initially i thought it was due to the changes in the PR, but it turns out that this is a known flakey test. I found that not closing the `bctx1` caused it to fail more often:

```diff
diff --git a/tests/browser_test.go b/tests/browser_test.go
index 9ec03f8..400f7f6 100644
--- a/tests/browser_test.go
+++ b/tests/browser_test.go
@@ -251,7 +251,7 @@ func TestMultiConnectToSingleBrowser(t *testing.T) {
 
        err = p1.Close(nil)
        require.NoError(t, err, "failed to close page #1")
-       bctx1.Close()
+       // bctx1.Close()
 
        p2, err := bctx2.NewPage()
        require.NoError(t, err, "failed to create page #2")
```

@inancgumus did some debugging and found that adding a sleep helped prevent the issue, so it's likely to be a race condition:

_"This works all the time (`-count=100`). This might mean that there is a race condition issue while creating pages."_

```go
// bctx1.Close()
time.Sleep(time.Second)    // <---
p2, err := bctx2.NewPage()
```

_"I made more checks. It's probably because of the browser or we killed the connection before attaching the blank page. And we're late updating/removing our internal contexts."_

```
making new context BB3319CCBC996F491E7F5E51F35D31E2
<- {"id":2,"result":{"windowId":1,"bounds":{"left":0,"top":397,"width":1280,"height":720,"windowState":"normal"}},"sessionId":"CF427B2EAF26043A97E976F6A0401CA7"}
<- {"id":23,"result":{},"sessionId":"09B020BCC266AC93E95C31D1FBAFA760"}
<- {"id":3,"result":{},"sessionId":"CF427B2EAF26043A97E976F6A0401CA7"}
<- {"method":"Target.detachedFromTarget","params":{"sessionId":"8EB506E362714513E689D093C0559848","targetId":"75D0B2C31C28330092C746AB234788AF"}}
<- {"method":"Target.detachedFromTarget","params":{"sessionId":"CF427B2EAF26043A97E976F6A0401CA7","targetId":"75D0B2C31C28330092C746AB234788AF"}}
<- {"method":"Target.detachedFromTarget","params":{"sessionId":"09B020BCC266AC93E95C31D1FBAFA760","targetId":"75D0B2C31C28330092C746AB234788AF"}}
newPageInContext BB3319CCBC996F491E7F5E51F35D31E2
<- {"id":4,"error":{"code":-32001,"message":"Session with given id not found."}}
```

1. Session `CF427B2EAF26043A97E976F6A0401CA7` gets closed before attaching to the page.
2. `Browser:initEvents:onDetachedFromTarget` happens before `newPageInContext`.

_"I guess the solution is we may want to include some mechanism to ensure the session with the target has not been detached before sending the get frame tree command (or others). Or, alternatively, we could handle the error and create a new session if necessary."_

**Comment by ankur22** [comment]
I have an inkling that the issue resides in chrome.

The error we get is from chrome which reads _"Session with given id not found."_. When creating a new page we work with the `Target.createTarget` CDP command, which only takes an optional browserContext id, and there is no reference of a session anywhere.

Sessions are created by chrome on `Target.createTarget`. The browser module doesn't execute any CDP commands to manage the sessions. Chrome asynchronously informs the browser module when to create or delete a session, so the browser module is beholden to chrome. If a sessions doesn't exist, it seems as though chrome has the race issue whereby it's unable to create the required session before it's needed.

It's also plausible that the `page.close` action causes side affects on all open browserContexts, but again, that's not something we have any control over.

Finally, maybe we have missed an important step when we're opening multiple websocket connections against chrome, maybe a flag of some kind. Having looked at puppeteer and playwright, I'm unable to see an obvious setup that would help prevent such an issue. In fact it seems that both playwright [hasn't solved it](https://github.com/microsoft/playwright/discussions/15735#discussioncomment-3209860) and [puppeteer](https://github.com/puppeteer/puppeteer/issues/9629) have encountered similar problems with unclear solutions.

I've opened an issue with the chromium team to see if they can help us out: https://bugs.chromium.org/p/chromium/issues/detail?id=1463459.

**Comment by ankur22** [comment]
After receiving a response from the chromium team with the [issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1463459) I raised a few weeks ago, they have come back with a response.

TL;DR

The `Session with given id not found. (-32001)` error is in fact for another request, and earlier request from a session which just happens to have the same message id as the request to `createTarget`. It seems that some error responses (at least this one) do not contain the `sessionId` that was in the original request:

```bash
msg='-> {"id":4,"sessionId":"169957DD1CBB5704DBC6798D891A6099","method":"Page.getFrameTree"}'
msg='<- {"method":"Target.detachedFromTarget","params":{"sessionId":"A10EDFD7D1B1CD6CFC1C8BBFCCFA2160","targetId":"6A497BD7D0C068592869FEEDE3235124"}}'
msg='<- {"method":"Target.detachedFromTarget","params":{"sessionId":"169957DD1CBB5704DBC6798D891A6099","targetId":"6A497BD7D0C068592869FEEDE3235124"}}'
msg='<- {"method":"Target.detachedFromTarget","params":{"sessionId":"60BE8201E7C7B6136367BB624399D03E","targetId":"6A497BD7D0C068592869FEEDE3235124"}}'
msg='-> {"id":4,"method":"Target.createTarget","params":{"url":"about:blank","browserContextId":"A42B57D4FDDF0A5E3928DC583BAD8A9E"}}'
msg='<- {"id":4,"error":{"code":-32001,"message":"Session with given id not found."}}'
```

There's the request for `Page.getFrameTree`, and then we have a request for `Target.createTarget`, which both have the same `id`. The error response is actually for the `Page.getFrameTree` request, but we redirect it to any handlers listening on the connection event loop (which is what `Target.createTarget` was sent on) because the response doesn't contain the `sessionId`.

The work around are:

1. Use unique Ids.
2. The chromium team suggest that it's ok to ignore `"code":-32001` errors, so we should do that.
3. When we receive an error message without a `sessionId`, it feels like we should be sending that response to all open `session`s and the `connection` event loop too. WDYT?

**Comment by ankur22** [comment]
Resolved in https://github.com/grafana/xk6-browser/pull/984.

**Comment by inancgumus** [comment]
Closed by #1219.
